File: 290.word-pattern.java
/*
 * @lc app=leetcode id=290 lang=java
 *
 * [290] Word Pattern
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");

        if (words.length != pattern.length()) {
            return false;
        }
        
        Map<Character, String> letterToWord = new HashMap<>();
        Map<String, Character> wordToLetter = new HashMap<>();

        for (int i = 0; i < words.length; i++) {
            String word = words[i];
            char letter = pattern.charAt(i);

            if (letterToWord.containsKey(letter)) {
                if (!letterToWord.get(letter).equals(word)) {
                    return false;
                }
            } else {
                letterToWord.put(letter, word);
            }

            if (wordToLetter.containsKey(word)) {
                if (!wordToLetter.get(word).equals(letter)) {
                    return false;
                }
            } else {
                wordToLetter.put(word, letter);
            }
        }

        return true;
    }
}
// @lc code=end

File: 290.word-pattern.py
#
# @lc app=leetcode id=290 lang=python3
#
# [290] Word Pattern
#

# @lc code=start

class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(words) != len(pattern):
             return False

        pattern_to_word = {}
        word_to_pattern = {}

        for idx, letter in enumerate(pattern):
            word = words[idx]
        
            # Check pattern -> word mapping
            if letter in pattern_to_word:
                if pattern_to_word[letter] != word:
                    return False
            else:
                pattern_to_word[letter] = word

            # Check word -> pattern mapping
            if word in word_to_pattern:
                if word_to_pattern[word] != letter:
                    return False
            else:
                word_to_pattern[word] = letter
    
        
        return True
        # s = s.split()

        # return (len(set(pattern)) ==
        #         len(set(s)) ==
        #         len(set(zip_longest(pattern,s))))
        
        
# @lc code=end

